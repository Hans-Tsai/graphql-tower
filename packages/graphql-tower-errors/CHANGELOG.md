# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="3.21.3"></a>
## [3.21.3](https://github.com/tmotx/graphql-tower/compare/v3.21.2...v3.21.3) (2017-12-05)




**Note:** Version bump only for package graphql-tower-errors

<a name="3.16.0"></a>
# [3.16.0](https://github.com/tmotx/graphql-tower/compare/v3.15.1...v3.16.0) (2017-11-19)


### Features

* **help:** added assertResult ([e6cd019](https://github.com/tmotx/graphql-tower/commit/e6cd019))




<a name="3.15.1"></a>
## [3.15.1](https://github.com/tmotx/graphql-tower/compare/v3.15.0...v3.15.1) (2017-11-18)


### Bug Fixes

* **error:** added NotAcceptableError ([7258d43](https://github.com/tmotx/graphql-tower/commit/7258d43))
