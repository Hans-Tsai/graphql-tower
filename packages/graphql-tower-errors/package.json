{
  "name": "graphql-tower-errors",
  "version": "1.0.0",
  "description": "",
  "main": "lib/index.js",
  "scripts": {
    "build": "rm -rf lib/* && babel src --ignore __tests__ --out-dir lib"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tmotx/graphql-tower.git"
  },
  "author": "tmotx",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/tmotx/graphql-tower/issues"
  },
  "homepage": "https://github.com/tmotx/graphql-tower#readme",
  "files": [
    "lib"
  ],
  "peerDependencies": {
    "graphql": "^0.11.7"
  },
  "devDependencies": {
    "faker": "^4.1.0"
  }
}
