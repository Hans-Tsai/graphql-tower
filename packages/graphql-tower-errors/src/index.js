import { GraphQLError } from 'graphql/error';

export class ForbiddenError extends GraphQLError {
  symbol = 'ForbiddenError';

  constructor(message = 'ForbiddenError') {
    super(message);
  }
}

export class NotFoundError extends GraphQLError {
  symbol = 'NotFoundError';

  constructor(message = 'NotFoundError') {
    super(message);
  }
}

export class UnauthorizedError extends GraphQLError {
  symbol = 'UnauthorizedError';

  constructor(message = 'UnauthorizedError') {
    super(message);
  }
}

export class PaymentRequiredError extends GraphQLError {
  symbol = 'PaymentRequiredError';

  constructor(message = 'PaymentRequiredError') {
    super(message);
  }
}

export class NotAcceptableError extends GraphQLError {
  symbol = 'NotAcceptableError';

  constructor(message = 'NotAcceptableError') {
    super(message);
  }
}

export class GoneDataError extends GraphQLError {
  symbol = 'GoneDataError';

  constructor(message = 'GoneDataError') {
    super(message);
  }
}

export class ConflictError extends GraphQLError {
  symbol = 'ConflictError';

  constructor(message = 'ConflictError') {
    super(message);
  }
}

export class LockedError extends GraphQLError {
  symbol = 'LockedError';

  constructor(message = 'LockedError') {
    super(message);
  }
}

export class UnavailableForLegalReasonsError extends GraphQLError {
  symbol = 'UnavailableForLegalReasonsError';

  constructor(message = 'UnavailableForLegalReasonsError') {
    super(message);
  }
}
