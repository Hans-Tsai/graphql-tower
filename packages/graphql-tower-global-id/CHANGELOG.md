# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="3.39.10"></a>
## [3.39.10](https://github.com/tmotx/graphql-tower/compare/v3.39.9...v3.39.10) (2018-08-15)


### Bug Fixes

* **all:** update packages ([e37ac1a](https://github.com/tmotx/graphql-tower/commit/e37ac1a))





<a name="3.3.3"></a>
## [3.3.3](https://github.com/tmotx/graphql-tower/compare/v3.3.2...v3.3.3) (2017-10-27)


### Bug Fixes

* **cache:** fix get model from cache will load model not data ([d7903b7](https://github.com/tmotx/graphql-tower/commit/d7903b7))




<a name="3.1.0"></a>
# [3.1.0](https://github.com/tmotx/graphql-tower/compare/v3.0.1...v3.1.0) (2017-10-26)


### Bug Fixes

* **tower:** fix export default ([12f42ca](https://github.com/tmotx/graphql-tower/commit/12f42ca))
