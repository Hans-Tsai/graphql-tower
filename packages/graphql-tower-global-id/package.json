{
  "name": "graphql-tower-global-id",
  "version": "1.0.0",
  "description": "",
  "main": "lib/index.js",
  "scripts": {
    "build": "rm -rf lib/* && babel src --ignore __tests__ --out-dir lib"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/tmotx/graphql-tower.git"
  },
  "author": "tmotx",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/tmotx/graphql-tower/issues"
  },
  "homepage": "https://github.com/tmotx/graphql-tower#readme",
  "files": [
    "lib"
  ],
  "dependencies": {
    "base-x": "^3.0.2",
    "bignumber.js": "^4.1.0",
    "lodash": "^4.17.4"
  }
}
