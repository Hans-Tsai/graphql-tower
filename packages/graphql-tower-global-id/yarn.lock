# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


base-x@^3.0.2:
  version "3.0.2"
  resolved "https://registry.yarnpkg.com/base-x/-/base-x-3.0.2.tgz#bf873861b7514279b7969f340929eab87c11d130"
  dependencies:
    safe-buffer "^5.0.1"

bignumber.js@^4.1.0:
  version "4.1.0"
  resolved "https://registry.yarnpkg.com/bignumber.js/-/bignumber.js-4.1.0.tgz#db6f14067c140bd46624815a7916c92d9b6c24b1"

lodash@^4.17.4:
  version "4.17.4"
  resolved "https://registry.yarnpkg.com/lodash/-/lodash-4.17.4.tgz#78203a4d1c328ae1d86dca6460e369b57f4055ae"

safe-buffer@^5.0.1:
  version "5.1.1"
  resolved "https://registry.yarnpkg.com/safe-buffer/-/safe-buffer-5.1.1.tgz#893312af69b2123def71f57889001671eeb2c853"
