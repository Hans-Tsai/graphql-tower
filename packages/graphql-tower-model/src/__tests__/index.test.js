import knex, { client } from 'jest-mock-knex';
import _ from 'lodash';
import faker from 'faker';
import { toGlobalId, fromGlobalId } from 'graphql-tower-global-id';
import { NotFoundError } from 'graphql-tower-errors';
import Model, { ValueColumn, HashColumn } from '../';

process.setMaxListeners(0);

const database = knex({
  client: 'pg',
  connection: {
    host: '127.0.0.1',
    user: 'postgres',
    password: null,
    database: 'graphql_tower',
  },
});


class DefaultModel extends Model {
  static database = database;

  static tableName = 'default_table';

  static hasOperator = true;

  static toKeyword = data => data.name;

  static columns = {
    name: new ValueColumn(),
    password: new HashColumn(),
    data: new ValueColumn(Object),
  }
}

class NoOperatorModel extends Model {
  static database = database;

  static tableName = 'default_table';

  static hasTimestamps = false;

  static columns = {
    name: new ValueColumn(),
  }
}


describe('model', () => {
  beforeAll(async () => {
    await database.schema.dropTableIfExists('default_table');

    await database.schema.createTable('default_table', (table) => {
      table.increments();
      table.string('name');
      table.text('password');
      table.specificType('keyword', 'tsvector');
      table.jsonb('data');
      table.timestamps();
      table.integer('created_by');
      table.integer('updated_by');
      table.timestamp('deleted_at');
      table.integer('deleted_by');
    });

    await database('default_table').insert({ name: 'name is one' });
    await database('default_table').insert({ name: 'name is two' });
  });

  afterAll(() => database.destroy());

  describe('static', () => {
    describe('columns', () => {
      it('object', () => {
        expect(DefaultModel.columns).toMatchSnapshot();
      });

      it('thunk', () => {
        class ThunkModel extends Model {
          static database = database;

          static tableName = 'default_table';

          static columns = () => ({
            name: new ValueColumn(),
            password: new HashColumn(),
            data: new ValueColumn(Object),
          })
        }

        expect(ThunkModel.columns).toMatchSnapshot();
      });
    });

    it('dataloader', () => {
      expect(DefaultModel.dataloader).toMatchSnapshot();
    });

    it('queryBuilder', () => {
      expect(DefaultModel.queryBuilder.toString()).toBe('select * from "default_table"');
    });

    describe('toGlobalId & fromGlobalId', () => {
      it('integer', () => {
        expect(DefaultModel.toGlobalId('2020')).toBe('iN3wDu0PJew8f0Jmr3hvaGDUPVMrPibs');
        expect(DefaultModel.fromGlobalId('iN3wDu0PJew8f0Jmr3hvaGDUPVMrPibs')).toBe('2020');

        expect(() => DefaultModel.fromGlobalId(toGlobalId('type', '2020'))).toThrowErrorMatchingSnapshot();
        expect(fromGlobalId(DefaultModel.toGlobalId('2020')).type).toBe('default_table');
      });

      it('uuid', () => {
        class UUIDModel extends Model {
          static database = database;

          static tableName = 'default_table';

          static hasUUID = true;

          static columns = {
            name: new ValueColumn(),
          }
        }

        expect(UUIDModel.toGlobalId('131d069a-8b6e-45d1-af3b-c25c598e06be')).toBe('iU1OZjz0bCBYpXCysIj8ELxRnJZZZRdhoTS1EaGgcve');
        expect(UUIDModel.fromGlobalId('iU1OZjz0bCBYpXCysIj8ELxRnJZZZRdhoTS1EaGgcve')).toBe('131d069a-8b6e-45d1-af3b-c25c598e06be');

        expect(() => UUIDModel.fromGlobalId(toGlobalId('default_table', '2020'))).toThrowErrorMatchingSnapshot();
      });
    });

    it('format && signify', () => {
      expect(DefaultModel.format({ is_admin: true })).toEqual({ isAdmin: true });
      expect(DefaultModel.signify({ isAdmin: true })).toEqual({ is_admin: true });
    });

    it('forge', () => {
      expect(DefaultModel.forge({ name: 'my name' })).toMatchSnapshot();
    });

    describe('load && loadMany', () => {
      it('softDelete', async () => {
        expect(await DefaultModel.load()).toBe(null);

        expect(await DefaultModel.load(1)).toEqual(expect.objectContaining({ name: 'name is one' }));

        const results = await Promise.all([DefaultModel.load(1), DefaultModel.loadMany([1, 2])]);
        expect(results).toEqual([
          expect.objectContaining({ name: 'name is one' }),
          [expect.objectContaining({ name: 'name is one' }), expect.objectContaining({ name: 'name is two' })],
        ]);

        expect(client).toMatchSnapshot();

        await expect(DefaultModel.load(99))
          .resolves.toEqual(null);
        await expect(DefaultModel.load(99, NotFoundError))
          .rejects.toEqual(new NotFoundError());
        await expect(DefaultModel.loadMany([99]))
          .resolves.toEqual([null]);
        await expect(DefaultModel.loadMany([99], NotFoundError))
          .rejects.toEqual(new NotFoundError());
      });

      it('no soft delete', async () => {
        class DirectDelete extends NoOperatorModel {
          static softDelete = false;
        }
        const model = new DefaultModel({ name: 'for no soft delete load' });
        await (await model.save(10)).destroy(10);
        expect(await DirectDelete.load(3))
          .toEqual(expect.objectContaining({ name: 'for no soft delete load' }));
        expect(client).toMatchSnapshot();
      });
    });

    it('isUUID', () => {
      expect(DefaultModel.isUUID(faker.random.uuid())).toBe(true);
      expect(DefaultModel.isUUID('XYZ')).toBe(false);
    });

    it('fromModel', () => {
      expect(DefaultModel.fromModel('20')).toBe('20');
      expect(DefaultModel.fromModel({ nativeId: '20' })).toBe('20');
      expect(DefaultModel.fromModel(null)).toBe(null);
    });
  });

  describe('model', () => {
    it('isNew', () => {
      const model = new DefaultModel();
      expect(model.isNew).toBe(true);

      model.id = '20';
      expect(model.isNew).toBe(false);
    });

    it('changes', async () => {
      const model = await DefaultModel.load(1);
      expect(model.changes).toEqual({});

      model.name = 'a new name';
      expect(model.changes).toEqual({ name: 'a new name' });
    });

    it('query', async () => {
      const model = new DefaultModel();
      expect(model.queryBuilder.toString()).toBe('select * from "default_table"');
      expect(model.query.toString()).toBe('select * from "default_table" where "deleted_at" is null');
    });

    it('nativeId', async () => {
      const model = await DefaultModel.load(1);
      expect(model.id).toBe('iN3wDu0PJew8f0Jmr3hvaGDUPVMrPi5J');
      expect(model.nativeId).toBe(1);
    });

    describe('forge & valueOf', () => {
      it('from object', () => {
        const data = { name: 'a name' };
        const model = (new DefaultModel()).forge(data);
        expect(model.valueOf()).not.toBe(data);
        expect(model._.previous).toBe(data);
      });

      it('from model', async () => {
        const model = await DefaultModel.load(1);
        const other = (new DefaultModel()).forge(model);
        expect(other.valueOf()).not.toBe(model.valueOf());
        expect(other._.previous).toBe(model._.previous);
      });
    });

    it('clone', async () => {
      const model = await DefaultModel.load(1);
      model.name = 'new model one';

      const newModel = model.clone();
      expect(model).toEqual(newModel);
      expect(model._.previous).toBe(newModel._.previous);
      expect(model._.current).not.toBe(newModel._.current);

      newModel.name = 'new model clone';
      expect(model.name).toBe('new model one');
      expect(model.valueOf()).not.toEqual(newModel.valueOf());
    });

    it('merge', async () => {
      const model = await DefaultModel.load(1);
      model.merge({ name: 'a name', archive: { items: [20, 40] } });
      expect(model._).toMatchSnapshot();
    });

    it('hash & verify', async () => {
      const model = await DefaultModel.load(1);
      model.set({ password: 'XYZ2020' });
      expect(model.verify('password', 'XYZ2020')).toBe(true);
      expect(model.verify('password', 'XYZ2049')).toBe(false);
    });

    describe('insert of save', () => {
      it('has operator', async () => {
        await (new DefaultModel({ name: 'new is insert' })).save(10);
        await (new DefaultModel()).save(10, { name: 'new is action' });
        expect(client).toMatchSnapshot();

        await expect((new DefaultModel()).save(null, { name: 'new is insert again' }))
          .rejects.toEqual(new Error('operator is required'));
      });

      it('no operator', async () => {
        await (new NoOperatorModel()).save(null, { name: 'new is no operator insert' });
        expect(client).toMatchSnapshot();
      });
    });

    describe('update of save', async () => {
      it('has operator', async () => {
        const model = await DefaultModel.load(4);

        await model.save(20);

        model.name = 'name is update';
        await model.save(20);

        await model.save(20, { name: 'name is action' });
        expect(client).toMatchSnapshot();

        await expect(model.save(null, { name: 'new is update again' }))
          .rejects.toEqual(new Error('operator is required'));
      });

      it('no operator', async () => {
        const model = await NoOperatorModel.load(6);
        await model.save(null, { name: 'name is no operator update' });
        expect(client).toMatchSnapshot();
      });
    });

    describe('destroy', () => {
      it('has operator', async () => {
        const model = await DefaultModel.load(4);

        await expect(model.destroy())
          .rejects.toEqual(new Error('operator is required'));

        await model.destroy(10);

        await model.destroy(10);
        expect(client).toMatchSnapshot();
      });

      it('no operator', async () => {
        const model = await NoOperatorModel.load(5);
        await model.destroy();
        expect(client).toMatchSnapshot();
      });

      it('is not soft delete', async () => {
        class DirectDelete extends NoOperatorModel {
          static softDelete = false;
        }
        const model = await DirectDelete.load(6);

        await model.destroy();
        expect(client).toMatchSnapshot();
      });
    });

    it('fetch', async () => {
      const model = new DefaultModel({ name: 'name is one' });
      await model.fetch();
      expect(model.valueOf()).toMatchSnapshot();
      expect(client).toMatchSnapshot();

      await expect((new DefaultModel({ name: 'one' })).fetch())
        .resolves.toEqual(null);
      await expect((new DefaultModel({ name: 'one' })).fetch(NotFoundError))
        .rejects.toEqual(new NotFoundError());
    });

    it('fetch with where', async () => {
      const model = new DefaultModel();
      model.where({ name: 'name is one' });
      await model.fetch();
      expect(model.valueOf()).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });

    describe('fetchAll', () => {
      it('no cache', async () => {
        const model = new DefaultModel({ name: 'name is one' });
        expect(_.map(await model.fetchAll(), data => data.name)).toMatchSnapshot();
        expect(client).toMatchSnapshot();

        await expect((new DefaultModel({ name: 'one' })).fetchAll())
          .resolves.toEqual([]);
        await expect((new DefaultModel({ name: 'one' })).fetchAll(NotFoundError))
          .rejects.toEqual(new NotFoundError());
      });

      it('cache', async () => {
        const model = new DefaultModel({ name: 'name is one' });
        model.cache = { prime: jest.fn() };
        await model.fetchAll();
        expect(model.cache.prime).toHaveBeenCalledWith(toGlobalId('default_table', 1), expect.anything());
      });
    });

    it('addKeyValue', async () => {
      const model = await DefaultModel.load(1);

      expect(model.valueOf('data')).toBe(null);
      await model.addKeyValue('data', '10', 'xyz');
      expect(model.valueOf('data')).toEqual({ 10: 'xyz' });

      expect(client).toMatchSnapshot();
    });

    it('delKeyValue', async () => {
      const model = await DefaultModel.load(1);

      expect(model.valueOf('data')).toEqual({ 10: 'xyz' });
      await model.delKeyValue('data', '10');
      expect(model.valueOf('data')).toEqual({});

      expect(client).toMatchSnapshot();
    });

    it('search', async () => {
      await (new DefaultModel({ name: 'new for action' })).save('10');
      await (new DefaultModel({ name: 'new for search' })).save('10');
      const model = new DefaultModel();
      model.search('new');
      expect(_.map(await model.fetchAll(), data => data.name)).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });
  });
});
