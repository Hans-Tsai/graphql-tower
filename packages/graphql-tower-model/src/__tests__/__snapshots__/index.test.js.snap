// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`model model addKeyValue 1`] = `
Array [
  "select * from default_table where id in (1) and deleted_at is null",
  "update default_table set data = coalesce(data, '{}') || {\\"10\\":\\"xyz\\"} where deleted_at is null and id = 1",
]
`;

exports[`model model delKeyValue 1`] = `
Array [
  "select * from default_table where id in (1) and deleted_at is null",
  "update default_table set data = coalesce(data, '{}') - 10 where deleted_at is null and id = 1",
]
`;

exports[`model model destroy has operator 1`] = `
Array [
  "select * from default_table where id in (4) and deleted_at is null",
  "update default_table set deleted_at = DATE, deleted_by = 10 where deleted_at is null and id = 4",
]
`;

exports[`model model destroy is not soft delete 1`] = `
Array [
  "select * from default_table where id in (6) and deleted_at is null",
  "delete from default_table where id = 6",
]
`;

exports[`model model destroy no operator 1`] = `
Array [
  "select * from default_table where id in (5) and deleted_at is null",
  "update default_table set deleted_at = DATE where deleted_at is null and id = 5",
]
`;

exports[`model model fetch 1`] = `
Object {
  "createdAt": null,
  "createdBy": null,
  "data": null,
  "deletedAt": null,
  "deletedBy": null,
  "id": 1,
  "keyword": null,
  "name": "name is one",
  "password": null,
  "updatedAt": null,
  "updatedBy": null,
}
`;

exports[`model model fetch 2`] = `
Array [
  "select * from default_table where deleted_at is null and name = name is one limit 1",
]
`;

exports[`model model fetch with where 1`] = `
Object {
  "createdAt": null,
  "createdBy": null,
  "data": null,
  "deletedAt": null,
  "deletedBy": null,
  "id": 1,
  "keyword": null,
  "name": "name is one",
  "password": null,
  "updatedAt": null,
  "updatedBy": null,
}
`;

exports[`model model fetch with where 2`] = `
Array [
  "select * from default_table where name = name is one and deleted_at is null limit 1",
]
`;

exports[`model model fetchAll no cache 1`] = `
Array [
  "name is one",
]
`;

exports[`model model fetchAll no cache 2`] = `
Array [
  "select * from default_table where deleted_at is null and name = name is one limit 1000",
]
`;

exports[`model model insert of save has operator 1`] = `
Array [
  "insert into default_table (created_at, created_by, keyword, name, updated_at, updated_by) values (DATE, 10, new is insert, new is insert, DATE, 10) returning id",
  "insert into default_table (created_at, created_by, keyword, name, updated_at, updated_by) values (DATE, 10, new is action, new is action, DATE, 10) returning id",
]
`;

exports[`model model insert of save no operator 1`] = `
Array [
  "insert into default_table (name) values (new is no operator insert) returning id",
]
`;

exports[`model model merge 1`] = `
Object {
  "current": Object {
    "archive": Object {
      "items": Array [
        20,
        40,
      ],
    },
    "createdAt": null,
    "createdBy": null,
    "data": null,
    "deletedAt": null,
    "deletedBy": null,
    "id": 1,
    "keyword": null,
    "name": "a name",
    "password": null,
    "updatedAt": null,
    "updatedBy": null,
  },
  "previous": Object {
    "archive": Object {
      "items": Array [
        20,
        40,
      ],
    },
    "createdAt": null,
    "createdBy": null,
    "data": null,
    "deletedAt": null,
    "deletedBy": null,
    "id": 1,
    "keyword": null,
    "name": "a name",
    "password": null,
    "updatedAt": null,
    "updatedBy": null,
  },
  "queryBuilder": undefined,
}
`;

exports[`model model search 1`] = `
Array [
  "new for action",
  "new for search",
]
`;

exports[`model model search 2`] = `
Array [
  "insert into default_table (created_at, created_by, keyword, name, updated_at, updated_by) values (DATE, 10, new for action, new for action, DATE, 10) returning id",
  "insert into default_table (created_at, created_by, keyword, name, updated_at, updated_by) values (DATE, 10, new for search, new for search, DATE, 10) returning id",
  "select *, ts_rank(keyword, new) as rank from default_table where keyword @@ to_tsquery(new) and deleted_at is null order by rank desc limit 1000",
]
`;

exports[`model model update of save has operator 1`] = `
Array [
  "select * from default_table where id in (4) and deleted_at is null",
  "update default_table set keyword = name is update, name = name is update, updated_at = DATE, updated_by = 20 where deleted_at is null and id = 4",
  "update default_table set keyword = name is action, name = name is action, updated_at = DATE, updated_by = 20 where deleted_at is null and id = 4",
]
`;

exports[`model model update of save no operator 1`] = `
Array [
  "select * from default_table where id in (6) and deleted_at is null",
  "update default_table set keyword = null, name = name is no operator update, updated_at = null, updated_by = null where deleted_at is null and id = 6",
]
`;

exports[`model static columns object 1`] = `
Object {
  "data": ValueColumn {
    "_": Object {
      "name": undefined,
      "serialize": [Function],
    },
    "enumerable": true,
  },
  "name": ValueColumn {
    "_": Object {
      "name": undefined,
      "serialize": [Function],
    },
    "enumerable": true,
  },
  "password": HashColumn {
    "_": Object {
      "name": undefined,
      "serialize": [Function],
    },
    "enumerable": true,
  },
}
`;

exports[`model static columns thunk 1`] = `
Object {
  "data": ValueColumn {
    "_": Object {
      "name": undefined,
      "serialize": [Function],
    },
    "enumerable": true,
  },
  "name": ValueColumn {
    "_": Object {
      "name": undefined,
      "serialize": [Function],
    },
    "enumerable": true,
  },
  "password": HashColumn {
    "_": Object {
      "name": undefined,
      "serialize": [Function],
    },
    "enumerable": true,
  },
}
`;

exports[`model static dataloader 1`] = `
DataLoader {
  "_batchLoadFn": [Function],
  "_options": Object {
    "cache": false,
  },
  "_promiseCache": Map {},
  "_queue": Array [],
}
`;

exports[`model static forge 1`] = `
DefaultModel {
  "_": Object {
    "current": Object {
      "name": "my name",
    },
    "previous": Object {
      "name": "my name",
    },
    "queryBuilder": undefined,
  },
  "cache": null,
  "createdAt": null,
  "createdBy": null,
  "data": null,
  "deletedAt": null,
  "deletedBy": null,
  "id": null,
  "name": "my name",
  "password": null,
  "updatedAt": null,
  "updatedBy": null,
}
`;

exports[`model static load && loadMany no soft delete 1`] = `
Array [
  "insert into default_table (created_at, created_by, keyword, name, updated_at, updated_by) values (DATE, 10, for no soft delete load, for no soft delete load, DATE, 10) returning id",
  "update default_table set deleted_at = DATE, deleted_by = 10 where deleted_at is null and id = 3",
  "select * from default_table where id in (3)",
]
`;

exports[`model static load && loadMany softDelete 1`] = `
Array [
  "select * from default_table where id in (1) and deleted_at is null",
  "select * from default_table where id in (1, 2, 1) and deleted_at is null",
]
`;

exports[`model static toGlobalId & fromGlobalId integer 1`] = `"invalid global id"`;

exports[`model static toGlobalId & fromGlobalId uuid 1`] = `""`;
